overlord/fdestate/dbx.go

```diff
@@ -0,0 +1,568 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+package fdestate
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/gadget/device"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/fdestate/backend"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+var (
+	backendResealKeysForSignaturesDBUpdate = backend.ResealKeysForSignaturesDBUpdate
+)
+
+type EFISecurebootKeyDatabase int
+
+const (
+	EFISecurebootPK EFISecurebootKeyDatabase = iota
+	EFISecurebootKEK
+	EFISecurebootDB
+	EFISecurebootDBX
+)
+
+// EFISecureBootDBUpdatePrepare notifies notifies that the local EFI key
+// database manager is about to update the database.
+func EFISecureBootDBUpdatePrepare(st *state.State, db EFISecurebootKeyDatabase, payload []byte) error {
+	method, err := device.SealedKeysMethod(dirs.GlobalRootDir)
+	if err != nil {
+		if err == device.ErrNoSealedKeys {
+			return nil
+		}
+		return err
+	}
+
+	st.Lock()
+	defer st.Unlock()
+
+	if err := checkDBXChangeConflicts(st); err != nil {
+		return err
+	}
+
+	op, err := addEFISecurebootDBUpdateChange(st, method, payload)
+	if err != nil {
+		return err
+	}
+
+	chgID := op.ChangeID
+
+	chg := st.Change(chgID)
+
+	// we're good so far, kick off the change
+	st.EnsureBefore(0)
+
+	// we want to synchronize with the prepare task completing successfully as
+	// at this point the keys will have been resealed with the proposed DBX
+	// payload
+	chgFailed := false
+	st.Unlock()
+	afterPrepareOKC := awaitDBXUpdatePrepareOK(st, chgID)
+	select {
+	case <-afterPrepareOKC:
+		// prepare step has completed successfully
+	case <-chg.Ready():
+		// change failed unexpectedly
+		chgFailed = true
```

Comment:
do we need some form of timeout here?

Reply:
I don't think so. <-chg.Ready() is only to handle an event where the change completes unexpectedly, caused by an error.

It's not clear what would be right course of action if the request was to timeout. We could in theory attempt to abort the change, but then it makes sense to followup waiting for the abort to complete, in which case the code would again be doing exact same wait, but in a different context. I'm open to suggestions.